src/workerd/api/headers.h

```diff
@@ -0,0 +1,295 @@
+                jsg::ByteString,
+                IteratorState<jsg::ByteString>,
+                keyOrValueIteratorNext)
+  JSG_ITERATOR(ValueIterator, values,
+                jsg::ByteString,
+                IteratorState<jsg::ByteString>,
+                keyOrValueIteratorNext)
+
+  // JavaScript API.
+
+  JSG_RESOURCE_TYPE(Headers, CompatibilityFlags::Reader flags) {
+    JSG_METHOD(get);
+    JSG_METHOD(getAll);
+    if (flags.getHttpHeadersGetSetCookie()) {
+      JSG_METHOD(getSetCookie);
+    }
+    JSG_METHOD(has);
+    JSG_METHOD(set);
+    JSG_METHOD(append);
+    JSG_METHOD_NAMED(delete, delete_);
+    JSG_METHOD(forEach);
+    JSG_METHOD(entries);
+    JSG_METHOD(keys);
+    JSG_METHOD(values);
+
+    JSG_INSPECT_PROPERTY(immutable, inspectImmutable);
+
+    JSG_ITERABLE(entries);
+
+    JSG_TS_DEFINE(type HeadersInit = Headers | Iterable<Iterable<string>> | Record<string, string>);
+    // All type aliases get inlined when exporting RTTI, but this type alias is included by
+    // the official TypeScript types, so users might be depending on it.
+
+    JSG_TS_OVERRIDE({
+      constructor(init?: HeadersInit);
+
+      entries(): IterableIterator<[key: string, value: string]>;
+      [Symbol.iterator](): IterableIterator<[key: string, value: string]>;
+
+      forEach<This = unknown>(callback: (this: This, value: string, key: string, parent: Headers) => void, thisArg?: This): void;
+    });
+  }
+
+  void serialize(jsg::Lock& js, jsg::Serializer& serializer);
+  static jsg::Ref<Headers> deserialize(
+      jsg::Lock& js, rpc::SerializationTag tag, jsg::Deserializer& deserializer);
+
+  JSG_SERIALIZABLE(rpc::SerializationTag::HEADERS);
+
+  void visitForMemoryInfo(jsg::MemoryTracker& tracker) const;
+
+private:
+  // The header key can be stored either as a common header ID (uint) or an an
+  // UncommonHeaderKey. This class encapsulates the lower-cased header name and
+  // precomputed hash code for efficient comparisons. It is used for any header
+  // name that is not in the common header list.
+  class UncommonHeaderKey final {
+   public:
+    UncommonHeaderKey(UncommonHeaderKey&&) = default;
+    UncommonHeaderKey& operator=(UncommonHeaderKey&&) = default;
+    KJ_DISALLOW_COPY(UncommonHeaderKey);
+
+    inline bool operator==(const UncommonHeaderKey& other) const;
+    inline bool operator==(kj::StringPtr otherName) const;
+
+    inline uint hashCode() const { return hash; }
+    inline kj::StringPtr getName() const { return name; }
+
+    JSG_MEMORY_INFO(UncommonHeaderKey) {
+      tracker.trackField("name", name);
+    }
+
+    inline UncommonHeaderKey clone() const {
+      return UncommonHeaderKey(kj::str(name), hash);
+    }
+
+   private:
+    // The name is expected to be stored in lower-case form, but we do not
+    // enforce that actually within the struct. It is the responsibility of
+    // the caller to ensure it is appropriately lower-cased.
+    kj::String name;
+    kj::uint hash;
```

Comment:
KJ's hash table implementation already caches hash values for entries in the table, so caching it yourself is not necessary.